Alerting: Add experimental AlertRule and RecordingRule kinds

moustafab opened the pull request:
**What is this feature?**

This introduces a new app-platform app for rules (`rules.alerting.grafana.app`) which includes 2 new kinds:

- AlertRules - rules specifically for monitoring and alerting
- RecordingRules - rules used to record metrics based on another query or set of queries

In order to facilitate compatibility between the new APIs and the old APIs for rules we introduced a way to represent non-grouped rules as a group with a single rule with a sentinel mask as the group title. This sentinel mask is a formatted mask with the rule_uid included for uniqueness in the API, and would not be a valid Group title in the existing API.

Grouped rules in the new APIs can be read and manipulated, with some caveats:
- Grouped rules will have their group returned in a label `grafana.app/group` and the index in `grafana.app/group-index`
- Interval updates for grouped rules will cause updates to the other rules in the group

Non-grouped rules in the old APIs can be read and manipulated with some caveats:
- Non-grouped rules cannot have their group name changed
- Non-grouped rules cannot be grouped with other rules (only 1 rule allowed in the "group")
- Non-grouped rules cannot be created via the old APIs

NOTE: This is an experimental API still behind a feature-flag `kubernetesAlertingRules`.

**Why do we need this feature?**

These new APIs provide a way to manage rules as first class objects (without Groups), and are a step towards migrating fully to app-platform for alerting.

**Special notes for your reviewer:**

The definition of the kinds and the implementation were explicitly separated in commits to make it easier to review. The final commit contains some refactoring to reuse code around the list query so the prometheus api pagination and the new non-grouped rules work in a consistent way without code duplication.

Known limitations that are to be implemented in follow-up PRs:
- list filtering using label-selectors
- list filtering using field-selectors
- bulk deletes
- utility functions for dealing with prometheus `Duration` strings
- defaulting the error and nodata state enums properly via the api
- validation hooks in the app (moving validation logic up)
- making the UI representation of non-grouped rules more user-friendly

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

moustafab commented on pkg/registry/apps/alerting/rules/register.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+// GetAllowedV0Alpha1Resources returns the list of resources that are allowed to be accessed in v0alpha1.
+func (p *AlertingRulesAppInstaller) GetAllowedV0Alpha1Resources() []string {
+	return []string{
+		rulesv0alpha1.AlertRuleKind().Plural(),
+		rulesv0alpha1.RecordingRuleKind().Plural(),
+	}
+}
```
Probably a good place to wrap the feature flag?

moustafab committed:
Alerting: Define rules.alerting.grafana.app/v0alpha1 AlertRule and RecordingRule

The first version of the app platform apis for alerting rules, including AlertRule and RecordingRule definitions.

Co-authored-by: William Wernert <william.wernert@grafana.com>

moustafab committed:
Alerting: Add support for alpha rules apis in legacy storage

Rules created in the new api makes the rule have no group in the database, but the rule is returned in the old group api with a sentinel group name formatted with the rule uid for compatiblity with the old api.
This makes the UI continue to work with the rules without a group, and the ruler will continue to work with the rules without a group.

Rules are not allowed to be created in the provisioning api with a NoGroup sentinel mask, but NoGroup rules can be manipulated through both the new and old apis.

Co-authored-by: William Wernert <william.wernert@grafana.com>

yuri-tceretian reviewed (commented):
I took a look at schemas only

moustafab committed:
Alerting: Update prometheus api to reuse list query logic

This lets the prometheus api respect NoGroup query logic and treat non-grouped rules consistently.

Co-authored-by: William Wernert <william.wernert@grafana.com>

yuri-tceretian commented on pkg/registry/apps/alerting/rules/alertrule/compat.go:
```diff
@@ -0,0 +1,361 @@
+)
+
+func ConvertToK8sResource(
+	orgID int64,
```
Why don't we take orgID from rule definition?

yuri-tceretian reviewed (commented):
I am still reviewing this pull request. Please see my comments so far.
Also, there are no tests for changes in the regular API.

yuri-tceretian commented on pkg/registry/apps/alerting/rules/alertrule/compat.go:
```diff
@@ -0,0 +1,361 @@
+		meta.SetUpdatedBy(string(*rule.UpdatedBy))
+		k8sRule.SetUpdatedBy(string(*rule.UpdatedBy))
+	}
+	meta.SetUpdatedTimestamp(&rule.Updated)
+	k8sRule.SetUpdateTimestamp(rule.Updated)
```
Why do we need two annotations dedicated to updated timestamp?
It looks like 
```
      "grafana.app/updatedTimestamp": "2025-09-05T20:01:23Z",
      "grafana.com/updateTimestamp": "2025-09-05T16:01:23-04:00",
```

moustafab replied:
I don't really know, but they both existed and I didn't know which one to use.

yuri-tceretian commented on pkg/registry/apps/alerting/rules/alertrule/compat.go:
```diff
@@ -0,0 +1,361 @@
+	provenance := ngmodels.Provenance(k8sRule.GetProvenanceStatus())
+	if provenance == "" {
+		provenance = ngmodels.ProvenanceNone
+	}
```
Also, there is a bug. When I submit a rule with provisioned annotation "none", it sets it as provisioned

yuri-tceretian commented on apps/alerting/rules/kinds/v0alpha1/rule_spec.cue:
```diff
@@ -0,0 +1,46 @@
+	relativeTimeRange?: #RelativeTimeRange
+	datasourceUID:      #DatasourceUID
+	model:              _
+	source?:            bool
```
Also, there is a bug. It lets me create a rule with all source false

yuri-tceretian commented:
Also, there is no even basic validation of queries. I managed to create a query of a gibberish type "1234"

moustafab commented on apps/alerting/rules/kinds/v0alpha1/alertRule_spec.cue:
```diff
@@ -0,0 +1,39 @@
+	execErrState:                 ExecErrState
+	"for"?:                       string & #PromDuration
+	keepFiringFor?:               string & #PromDuration
+	missingSeriesEvalsToResolve?: int & >=0
```
The nesting doesn't help the user, and arguably makes things a bit more confusing. I can see how it might seem useful from our perspective that the values are grouped, but I think it makes the API more complicated.

rwwiv commented on pkg/registry/apps/alerting/rules/alertrule/authorize.go:
```diff
@@ -0,0 +1,50 @@
+	"github.com/grafana/grafana/pkg/services/accesscontrol"
+)
+
+func Authorize(ctx context.Context, ac accesscontrol.AccessControl, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
```
As above, we'll be checking folder permissions in a later stage.

yuri-tceretian commented on pkg/extensions/enterprise_imports.go:
```diff
@@ -53,5 +53,6 @@ import (
 	_ "github.com/grafana/e2e"
 	_ "github.com/grafana/gofpdf"
 	_ "github.com/grafana/gomemcache/memcache"
+	_ "github.com/grafana/grafana/apps/alerting/alertenrichment/pkg/apis/alertenrichment/v1beta1"
```
Is this relevant to the PR?

moustafab replied:
No, but somehow this was removed and needed to be re-added. I would guess it's an artifact of some bad merge and might be taken care of when we re-update and squash

alexander-akhmetov replied:
This is actually in the [latest main](https://github.com/grafana/grafana/blob/ebd1f63aab70387c1bf0bd9cfe4be9149abad015/pkg/extensions/enterprise_imports.go#L56). 🤔  But yes, recently someone removed it in an unrelated PR by accident, and I had to add it again.

yuri-tceretian commented on pkg/tests/apis/alerting/rules/compat/alertrule_test.go:
```diff
@@ -0,0 +1,421 @@
+			Annotations: map[string]string{
+				"grafana.app/folder": "test-folder",
+				// use provenance api to allow use of the provisioning api
+				"grafana.app/provenance": string(ngmodels.ProvenanceAPI),
```
we can use app extensions like SetProvenanceStatus etc so we can avoid sticking to annotation literals in our code.

yuri-tceretian commented on pkg/tests/apis/alerting/rules/compat/alertrule_test.go:
```diff
@@ -0,0 +1,421 @@
+	require.NoError(t, err)
+	require.NotNil(t, created)
+
+	t.Run("should be able to use the provisioning API with this rule", func(t *testing.T) {
```
Just a food for thoughts. I wonder if a better test could be defining full payload and expected responses  in files and then just create and compare JSONs. In other words, we create test-data files describe k8s resource in JSON and then expected responses from regular API and provisioning APIs. Then the test is just create a resource and compare responses of two APIs with those files.

yuri-tceretian commented on pkg/services/ngalert/provisioning/alert_rules.go:
```diff
@@ -242,10 +242,13 @@ func (service *AlertRuleService) GetAlertRuleWithFolderFullpath(ctx context.Cont
+// interval that is set in the rule struct and use the already existing group interval or the default one.
 func (service *AlertRuleService) CreateAlertRule(ctx context.Context, user identity.Requester, rule models.AlertRule, provenance models.Provenance) (models.AlertRule, error) {
+	if models.IsNoGroupRuleGroup(rule.RuleGroup) {
+		return models.AlertRule{}, fmt.Errorf("%w: rules must have a valid group", models.ErrAlertRuleFailedValidation)
```
I like error message below "rule group %s is reserved for no-group rules and cannot be used for rule groups with multiple rules". Let's use it here too?

yuri-tceretian replied:
Like "rule group %s is reserved for no-group rules and cannot be defined. Use empty group instead"

yuri-tceretian commented on pkg/services/ngalert/provisioning/alert_rules.go:
```diff
@@ -653,6 +663,9 @@ func (service *AlertRuleService) persistDelta(ctx context.Context, user identity
 				}
+				if models.IsNoGroupRuleGroup(update.Existing.RuleGroup) && !models.IsNoGroupRuleGroup(update.New.RuleGroup) {
+					return fmt.Errorf("%w: cannot move rule out of this group", models.ErrAlertRuleFailedValidation)
+				}
```
if I read it correctly, we do not allow moving a rule from "no-group" to a group. I am not sure that's correct.

yuri-tceretian reviewed (approved):
The new app code LGTM. My comments are non-blocking and may be addressed later. 
I believe that the logic around "no-group" group needs to be clearer, I am not sure I understand reason behind allowing users moving from a group to "no-group" but not letting them move it back to the group.

yuri-tceretian commented:
> Can you elaborate so I can get this into a test case? I believe technically query models can be basically gibberish as long as they're valid json strings.

In the regular API we do some basic validation of the query. The provisioning API may lack this validation

moustafab commented:
>  I believe that the logic around "no-group" group needs to be clearer, I am not sure I understand reason behind allowing users moving from a group to "no-group" but not letting them move it back to the group.

For posterity and to clarify:

We explicitly disallowed creating "ungrouped" rules via the old api and disallowed moving "ungrouped" rules (created in the new apis) to a group. This is intended to make migration from groups to no groups a one-way change, and simplify the logic around moving rules to a group to not have to consider no-group rules in the future. The compatibility between the two sets of APIs is necessary for UI functionality, but groups are going away in the new API and we want to limit their consideration in the future. This limitation is a first step in that direction.
